### ISSUE TITLE
Support for testing/mocking

### ISSUE THREAD

**Issue by owenconti** [hammerstonedev/sidecar#16]
It would be awesome if the package supported the ability to fake/mock a Sidecar execution, ie:

```php
/** @test */
public function it_calls_sidecar_function()
{
    // Given
    $mockedResponse = [
        'hello' => 'world'
    ];
    SomeFunction::fake($mockedResponse);

    // When
    $this->get('/call-function')->assertSuccessful();

    // Then
    $expectedParams = [
        'foo' => 'bar'
    ];
    SomeFunction::assertExecuted($expectedParams);
}
```

The idea here would be to keep the faking pattern similar to the Laravel first-party fakes (Http, Queue, etc):

- You call `fake()` with a mocked response for the lambda
- You call your controller/command/job, etc that will execute the lambda
- You call an assertion method to assert that the lambda was invoked with the given set of parameter

**Comment by aarondfrancis** [comment]
Yup this is a must. I like the general interface you've lined out here. 

I'll think on it, and also take a look at the Laravel Event and Bus fakes to see if there's something I can leverage there. I might be able to knock something out here in the next day or two, depending on how much the new kiddos sleep 😅

**Comment by owenconti** [comment]
No worries, I have a proof of concept on a fork that I can use in the meantime. Just working on one last piece and then I'll link it here.

**Comment by aarondfrancis** [comment]
That would be wonderful!

**Comment by owenconti** [comment]
Okay here it is: https://github.com/owenconti/sidecar/pull/1/files

Everything is new code except for the `isError` check. The mocked AWS response doesn't include the `FunctionError` key (I may be doing something wrong), which was causing the previous `isError` check to return true when it should have been false.

**Comment by aarondfrancis** [comment]
This is amazing. Thank you for doing this!

**Comment by owenconti** [comment]
No worries, hopefully it helps some!

**Comment by datashaman** [comment]
Since `Sidecar` extends `Facade`, it has the following functionality added which (imo) is a more useful mock than what you have here without any new code:

```
use Hammerstone\Sidecar\Sidecar;

Sidecar::shouldReceive('execute')
    ->once()
    ->with(MyFunction::class, 'bar')
    ->andReturn('results');
```

https://laravel.com/docs/9.x/mocking#mocking-facades

**Comment by mreduar** [comment]
> Since `Sidecar` extends `Facade`, it has the following functionality added which (imo) is a more useful mock than what you have here without any new code:
> 
> ```
> use Hammerstone\Sidecar\Sidecar;
> 
> // set the expectation in the test code
> Sidecar::shouldReceive('execute')
>     ->once()
>     ->with(SomeFunction::class, ['foo' => 'bar'])
>     ->andReturn('results');
> 
> // execute some runtime code that calls this internally
> SomeFunction::execute(['foo' => 'bar']);
> ```
> 
> https://laravel.com/docs/9.x/mocking#mocking-facades
> 
> Your method is also adding testing code in runtime classes. _Laravel_'s strategy with fakes is to encapsulate the testing code in a separate fake class and swap them out in the fake method. This ensures that the testing code affects the runtime code as little as possible.
> 
> Here's code for the Mail fake method as an example: https://github.com/laravel/framework/blob/9.x/src/Illuminate/Support/Facades/Mail.php#L45
> 
> But I don't believe it's necessary to create a fake class for just recording method executions. The mocking illustrated above does that already. Also read further in the docs about Facade spies.

I tried to do it as you indicated as I think it looks great, but I couldn't do it, I get an error about an incorrect number of parameters.
```
No matching handler found for Mockery_3_Hammerstone_Sidecar_Manager::execute('App\Sidecar\StorageDownloadZipped', ['files' => [...], 'tenant' => 'the-id-tenant'], false, 'RequestResponse'). Either the method was unexpected or its arguments matched no expected argument list for this method
```
